README.md
# Table of Contents

-   [Prologue](#org5a5b482)
-   [Packages](#org5162ade)
-   [Usage](#org8f95840)
-   [Development Process](#org500bff6)
    -   [Http Functions](#orgea60df6)
        -   [Get](#orgc53178d)
        -   [Post](#org4816e1a)
        -   [Bugs](#orgf07ea04)
            -   [Fix an error from inspect.Parameter which caused the function parallel down.](#org8cb169d):err:1:
    -   [Docs](#org471c09e)
        -   [Usage](#orgdd40ced)
-   [Epoligue](#org7d09938)
    -   [History](#org9b25ed6)
        -   [Version 1.0.2](#orgd045e62)
        -   [Version 1.0.1](#orgf937241)
        -   [Version 1.0.0](#org73249ca)
        -   [Version 0.1.2](#org872c4cb)
        -   [Version 0.1.1](#orga13def9)
        -   [Version 0.1.0](#org94f75db)



<a id="org5a5b482"></a>

# Prologue

Never had such a pure crawler like this `nacf`.

Although I often write crawlers, I don&rsquo;t like to use huge frameworks, such as scrapy, but prefer
simple `requests+bs4` or more general `requests_html`.  However, these two are inconvenient for a
crawler.  E.g. Places, such as error retrying or parallel crawling, need to be handwritten by
myself.  It is not very difficult to write it while writing too much can be tedious.  Hence I
started writing this nacf (Nasy Crawler Framework), hoping to simplify some error retrying or
parallel writing of crawlers.


<a id="org5162ade"></a>

# Packages

<table>
<caption class="t-above"><span class="table-number">Table 1:</span> Packages</caption>

<colgroup>
<col  class="org-left">

<col  class="org-right">

<col  class="org-left">
</colgroup>
<thead>
<tr>
<th scope="col" class="org-left">Package</th>
<th scope="col" class="org-right">Version</th>
<th scope="col" class="org-left">Description</th>
</tr>
</thead>

<tbody>
<tr>
<td class="org-left">requests-html</td>
<td class="org-right">0.10.0</td>
<td class="org-left">HTML Parsing for Humans.</td>
</tr>


<tr>
<td class="org-left">nalude</td>
<td class="org-right">0.3.0</td>
<td class="org-left">A standard module.  Inspired by Haskell&rsquo;s Prelude.</td>
</tr>
</tbody>
</table>


<a id="org8f95840"></a>

# Usage

see [tests](https://github.com/nasyxx/nacf/blob/master/tests/test_nacf.py).


<a id="org500bff6"></a>

# Development Process


<a id="orgea60df6"></a>

## DONE Http Functions

<p><span class="timestamp-wrapper"><span class="timestamp-kwd">CLOSED:</span> <span class="timestamp">&lt;Thu Feb 28 20:51:00 2019&gt;</span></span></p>


<a id="orgc53178d"></a>

### DONE Get

<p><span class="timestamp-wrapper"><span class="timestamp-kwd">CLOSED:</span> <span class="timestamp">&lt;Tue Dec 25 17:36:00 2018&gt;</span></span></p>


<a id="org4816e1a"></a>

### DONE Post

<p><span class="timestamp-wrapper"><span class="timestamp-kwd">CLOSED:</span> <span class="timestamp">&lt;Thu Feb 28 20:44:00 2019&gt;</span></span></p>


<a id="orgf07ea04"></a>

### DONE Bugs

<p><span class="timestamp-wrapper"><span class="timestamp-kwd">CLOSED:</span> <span class="timestamp">&lt;Thu Feb 28 20:51:00 2019&gt;</span></span></p>


<a id="org8cb169d"></a>

#### DONE Fix an error from inspect.Parameter which caused the function parallel down.     :err:1:

<p><span class="timestamp-wrapper"><span class="timestamp-kwd">CLOSED:</span> <span class="timestamp">&lt;Wed Dec 26 20:26:00 2018&gt;</span></span></p>


<a id="org471c09e"></a>

## NEXT Docs


<a id="orgdd40ced"></a>

### NEXT Usage


<a id="org7d09938"></a>

# Epoligue


<a id="org9b25ed6"></a>

## History


<a id="orgd045e62"></a>

### Version 1.0.2

-   **Data:** <span class="timestamp-wrapper"><span class="timestamp">&lt;Sun Mar 10, 2019&gt;</span></span>
-   **Changes:** Update nalude.


<a id="orgf937241"></a>

### Version 1.0.1

-   **Data:** <span class="timestamp-wrapper"><span class="timestamp">&lt;Sun Mar 10, 2019&gt;</span></span>
-   **Changes:** Update requests-html.


<a id="org73249ca"></a>

### Version 1.0.0

-   **Data:** <span class="timestamp-wrapper"><span class="timestamp">&lt;Thu Feb 28, 2019&gt;</span></span>
-   **Changes:** Now, old HTTP methods (`get` and `post`) cannot accept multiple URLs. Instead, we can use `gets` and `posts`.
-   **Adds:** -   `nacf.html`
    -   `nacf.json`
    -   `nacf.gets`
    -   `nacf.posts`
-   **Includes:** -   `nalude`


<a id="org872c4cb"></a>

### Version 0.1.2

-   **Data:** <span class="timestamp-wrapper"><span class="timestamp">&lt;Wed Dec 26, 2018&gt;</span></span>
-   **Fixed:** `inspect.Parameter` error in last version.


<a id="orga13def9"></a>

### Version 0.1.1

-   **Data:** <span class="timestamp-wrapper"><span class="timestamp">&lt;Wed Dec 26, 2018&gt;</span></span>
-   **Ignored:** An error caused by `inspect.Parameter`
-   **Help Wanted:** Can someone help me about the Parameter?


<a id="org94f75db"></a>

### Version 0.1.0

-   **Date:** <span class="timestamp-wrapper"><span class="timestamp">&lt;Sun Dec 23, 2018&gt;</span></span>
-   **Commemorate Version:** First Version
    -   Basic Functions.


CHANGELOG
* Version 1.0.3

+ Date :: <2019-05-09 Thu>
+ Changes :: Update requests-html

* Version 1.0.2

+ Date :: <2019-03-10 Sun>
+ Changes :: Update nalude.

* Version 1.0.1

+ Date :: <2019-03-10 Sun>
+ Changes :: Update requests-html.

* Version 1.0.0

+ Date :: <2019-02-28 Thu>
+ Changes :: Now, old HTTP methods (~get~ and ~post~) cannot accept multiple URLs. Instead, we can use ~gets~ and ~posts~.
+ Adds ::
    - ~nacf.html~
    - ~nacf.json~
    - ~nacf.gets~
    - ~nacf.posts~
+ Includes ::
    - ~nalude~

* Version 0.1.2

+ Date :: <2018-12-26 Wed>
+ Fixed :: ~inspect.Parameter~ error in last version.

* Version 0.1.1

+ Date :: <2018-12-26 Wed>
+ Ignored :: An error caused by ~inspect.Parameter~
+ Help Wanted :: Can someone help me about the Parameter?

* Version 0.1.0

+ Date :: <2018-12-23 Sun>
+ Commemorate Version :: First Version
    - Basic Functions.


tests/test_nacf.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
r"""
Life's pathetic, have fun ("▔□▔)/hi~♡ Nasy.

Excited without bugs::

    |             *         *
    |                  .                .
    |           .
    |     *                      ,
    |                   .
    |
    |                               *
    |          |\___/|
    |          )    -(             .              ·
    |         =\ -   /=
    |           )===(       *
    |          /   - \
    |          |-    |
    |         /   -   \     0.|.0
    |  NASY___\__( (__/_____(\=/)__+1s____________
    |  ______|____) )______|______|______|______|_
    |  ___|______( (____|______|______|______|____
    |  ______|____\_|______|______|______|______|_
    |  ___|______|______|______|______|______|____
    |  ______|______|______|______|______|______|_
    |  ___|______|______|______|______|______|____

author   : Nasy https://nasy.moe
date     : Dec 23, 2018
email    : Nasy <nasyxx+python@gmail.com>
filename : test_nacf.py
project  : tests
license  : LGPL-3.0+

There are more things in heaven and earth, Horatio, than are dreamt.
 --  From "Hamlet"
"""

# Standard Library
import unittest

# Prelude
from nalude import flatten

# Other Packages
from nacf import (css, get, gets, html, json, post, text,
                  urls, posts, xpath, parallel, __version__,)
from nacf.types import Res, Json, Iterable


class NacfTest(unittest.TestCase):
    """Nasy crawler framework test."""

    def test_version(self) -> None:
        """Test version of nacf."""
        with open("pyproject.toml") as f:
            for line in f:
                if "version" in line:
                    version = line.split()[-1].replace('"', "")
                    break
        self.assertEqual(__version__, version)

    def test_get_css(self) -> None:
        """Test get and css."""

        @get("python.org")
        @css(".widget-title", first=True)
        @text
        def crawler(res: str) -> str:
            """Test crawler."""
            return res

        self.assertEqual(crawler(), "Get Started")

    def test_gets_xpath(self) -> None:
        """Test gets and xpath."""

        @gets(["python.org", "python.org"])
        @xpath("//*[@class='widget-title']", first=True)
        @text
        def crawler(res: str) -> str:
            """Test crawler."""
            return res

        self.assertEqual(
            list(flatten(crawler())), ["Get Started", "Get Started"]
        )

    def test_post_json(self) -> None:
        """Test post, posts and json."""

        @post(
            "https://app.fakejson.com/q",
            json={
                "token": "FOk7RjbecxtWJHljGjCNjg",
                "data": {
                    "colorText": "colorText",
                    "colorHex": "colorHex",
                    "colorRGB": "colorRGB",
                    "colorHSL": "colorHSL",
                },
            },
        )
        @json
        def crawler(res: Json) -> Json:
            """Test crawler."""
            return res

        self.assertEqual(
            crawler(),
            {
                "colorText": "tufts blue",
                "colorRGB": "rgb(22, 75, 56)",
                "colorHex": "colorHex",
                "colorHSL": "hsl(233, 14%, 14%)",
            },
        )

    def test_posts(self) -> None:
        """Test posts and json."""

        @posts(
            ["https://app.fakejson.com/q"] * 3,
            jsons=[
                {
                    "token": "FOk7RjbecxtWJHljGjCNjg",
                    "data": {
                        "colorText": "colorText",
                        "colorHex": "colorHex",
                        "colorRGB": "colorRGB",
                        "colorHSL": "colorHSL",
                    },
                }
            ]
            * 3,
        )
        def crawler(res: Res) -> Iterable[Json]:
            """Test crawler."""
            return map(lambda r: r.json(), res)

        self.assertEqual(
            list(flatten(crawler())),
            [
                {
                    "colorText": "tufts blue",
                    "colorRGB": "rgb(22, 75, 56)",
                    "colorHex": "colorHex",
                    "colorHSL": "hsl(233, 14%, 14%)",
                }
            ]
            * 3,
        )

    def test_urls_text(self) -> None:
        """Test urls and text."""

        @urls(["python.org", "python.org"])
        @gets()
        @css(".widget-title", first=True)
        @text
        def crawler(res: str) -> str:
            """Test crawler."""
            return res

        self.assertEqual(
            list(flatten(crawler())), ["Get Started", "Get Started"]
        )

    def test_parallel_html(self) -> None:
        """Test parallel."""

        @urls(["python.org", "python.org"])
        @parallel()
        @get()
        @html
        def crawler(res: Res) -> str:
            """Test crawler."""
            return res.find(".widget-title", first=True).text

        self.assertEqual(
            list(flatten(crawler())), ["Get Started", "Get Started"]
        )


def run() -> None:
    """Run test."""
    unittest.main()


if __name__ == "__main__":
    run()


nacf/tools.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Life's pathetic, have fun ("▔□▔)/hi~♡ Nasy.

Excited without bugs::

    |             *         *
    |                  .                .
    |           .
    |     *                      ,
    |                   .
    |
    |                               *
    |          |\___/|
    |          )    -(             .              ·
    |         =\ -   /=
    |           )===(       *
    |          /   - \
    |          |-    |
    |         /   -   \     0.|.0
    |  NASY___\__( (__/_____(\=/)__+1s____________
    |  ______|____) )______|______|______|______|_
    |  ___|______( (____|______|______|______|____
    |  ______|____\_|______|______|______|______|_
    |  ___|______|______|______|______|______|____
    |  ______|______|______|______|______|______|_
    |  ___|______|______|______|______|______|____

author   : Nasy https://nasy.moe
date     : Dec 23, 2018
email    : Nasy <nasyxx+python@gmail.com>
filename : tools.py
project  : nacf
license  : LGPL-3.0+

There are more things in heaven and earth, Horatio, than are dreamt.
 --  From "Hamlet"
"""
# Standard Library
from inspect import Parameter, Signature, signature

# Prelude
from nalude import flatten

# Local Packages
from .types import Url, Func, Optional


def format_url(url: Optional[Url]) -> Optional[Url]:
    """Add schema to url."""
    return (
        url if url and url.startswith("http") else url and ("https://" + url)
    )


def generate_sig(func: Func, *params: Parameter) -> Signature:
    """Generate signature with param from func."""
    return signature(func).replace(
        parameters=list(signature(func).parameters.values()) + list(params)
    )


__all__ = ["format_url", "flatten"]


nacf/types.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Life's pathetic, have fun ("▔□▔)/hi~♡ Nasy.

Excited without bugs::

    |             *         *
    |                  .                .
    |           .
    |     *                      ,
    |                   .
    |
    |                               *
    |          |\___/|
    |          )    -(             .              ·
    |         =\ -   /=
    |           )===(       *
    |          /   - \
    |          |-    |
    |         /   -   \     0.|.0
    |  NASY___\__( (__/_____(\=/)__+1s____________
    |  ______|____) )______|______|______|______|_
    |  ___|______( (____|______|______|______|____
    |  ______|____\_|______|______|______|______|_
    |  ___|______|______|______|______|______|____
    |  ______|______|______|______|______|______|_
    |  ___|______|______|______|______|______|____

author   : Nasy https://nasy.moe
date     : Dec 23, 2018
email    : Nasy <nasyxx+python@gmail.com>
filename : types.py
project  : nacf
license  : LGPL-3.0+

There are more things in heaven and earth, Horatio, than are dreamt.
 --  From "Hamlet"
"""
# Standard Library
from typing import (IO, Dict, List, Tuple, Union, TypeVar,
                    Callable, Iterable, Optional, ByteString,)

# Web Packages
from requests_html import HTML
from requests_html import Element as Ele
from requests_html import HTMLResponse as Res

# Normal
a = TypeVar("a")
b = TypeVar("b")

# function
RT = TypeVar("RT")
Func = Callable[..., RT]

Reses = Iterable[Res]
RFunc = Callable[..., Res]
RsFunc = Callable[..., Reses]

Eles = Iterable[Ele]
EFunc = Callable[..., Union[Ele, Eles]]

SFunc = Callable[..., Union[str, Iterable[str]]]

# request
Data = Optional[Union[Dict[str, a], ByteString, IO]]
Json = Dict[str, a]
Url = str
Urls = Iterable[Url]

__all__ = [
    "Res",
    "Reses",
    "HTML",
    "a",
    "b",
    "RT",
    "Func",
    "RFunc",
    "RsFunc",
    "SFunc",
    "Data",
    "Json",
    "Url",
    "Urls",
    "List",
    "Tuple",
]


nacf/__init__.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
r"""
Life's pathetic, have fun ("▔□▔)/hi~♡ Nasy.

Excited without bugs::

    |             *         *
    |                  .                .
    |           .
    |     *                      ,
    |                   .
    |
    |                               *
    |          |\___/|
    |          )    -(             .              ·
    |         =\ -   /=
    |           )===(       *
    |          /   - \
    |          |-    |
    |         /   -   \     0.|.0
    |  NASY___\__( (__/_____(\=/)__+1s____________
    |  ______|____) )______|______|______|______|_
    |  ___|______( (____|______|______|______|____
    |  ______|____\_|______|______|______|______|_
    |  ___|______|______|______|______|______|____
    |  ______|______|______|______|______|______|_
    |  ___|______|______|______|______|______|____

author   : Nasy https://nasy.moe
date     : Dec 23, 2018
email    : Nasy <nasyxx+python@gmail.com>
filename : __init__.py
project  : nacf
license  : LGPL-3.0+

There are more things in heaven and earth, Horatio, than are dreamt.
 --  From "Hamlet"
"""
__version__ = "1.0.3"

# Local Packages
from .models import (css, get, gets, html, json, post,
                     text, urls, posts, xpath, parallel,)

if __name__ == "__main__":
    assert css
    assert get
    assert gets
    assert post
    assert posts
    assert html
    assert json
    assert text
    assert urls
    assert xpath
    assert parallel


nacf/models.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Life's pathetic, have fun ("▔□▔)/hi~♡ Nasy.

Excited without bugs::

    |             *         *
    |                  .                .
    |           .
    |     *                      ,
    |                   .
    |
    |                               *
    |          |\___/|
    |          )    -(             .              ·
    |         =\ -   /=
    |           )===(       *
    |          /   - \
    |          |-    |
    |         /   -   \     0.|.0
    |  NASY___\__( (__/_____(\=/)__+1s____________
    |  ______|____) )______|______|______|______|_
    |  ___|______( (____|______|______|______|____
    |  ______|____\_|______|______|______|______|_
    |  ___|______|______|______|______|______|____
    |  ______|______|______|______|______|______|_
    |  ___|______|______|______|______|______|____

author   : Nasy https://nasy.moe
date     : Dec 24, 2018
update   : Feb 28, 2019
email    : Nasy <nasyxx+python@gmail.com>
filename : models.py
project  : nacf
license  : LGPL-3.0+

There are more things in heaven and earth, Horatio, than are dreamt.
 --  From "Hamlet"
"""
# Standard Library
from inspect import Parameter
from functools import wraps
from itertools import chain, cycle
from multiprocessing.dummy import Pool

# Web Packages
from requests_html import HTMLSession as s

# Local Packages
from .tools import format_url, generate_sig
from .types import (RT, HTML, Ele, Res, Url, Data, Eles, Func, Json,
                    List, Urls, EFunc, Reses, RFunc, SFunc, Tuple,
                    Union, RsFunc, Callable, Iterable, Optional, a, b,)


def urls(urls_: Urls, **kwds: a) -> Callable[[Func], Func]:
    """Feed urls to next function."""

    def decorator(func: Func) -> Func:
        """Decorate function func."""

        @wraps(func)
        def wrapper() -> RT:
            """Wrap function func."""
            return func(_urls=urls_, **kwds)

        return wrapper

    return decorator


def parallel(n: Optional[int] = None, **kwds: a) -> Callable[[Func], Func]:
    """Run parallelly."""

    def decorator(func: Func) -> Func:
        """Decorate function func."""

        @wraps(func)
        def wrapper(
            _urls: Urls,
            *,
            _datas: Optional[Iterable[Data]] = None,
            _jsons: Optional[Iterable[Json]] = None,
            **_kwds: b,
        ) -> RT:
            """Wrap function func."""

            def runner(udj: Tuple[Url, Optional[Data], Optional[Json]]) -> RT:
                """Run func wrapper."""
                url, data, json = udj
                return (
                    func(url, _data=data, _json=json, **_kwds, **kwds)
                    if data or json
                    else func(url, **_kwds, **kwds)
                )

            with Pool(n) as pool:
                res = pool.map(
                    runner,
                    zip(
                        _urls,
                        _datas or cycle([None]),
                        _jsons or cycle([dict()]),
                    ),
                )
            return res

        setattr(
            wrapper,
            "__signature__",
            generate_sig(
                func,
                Parameter(
                    "_datas",
                    Parameter.KEYWORD_ONLY,
                    default=None,
                    annotation="Optional[Iterable[Data]]",
                ),
                Parameter(
                    "_jsons",
                    Parameter.KEYWORD_ONLY,
                    default=None,
                    annotation="Optional[Iterable[Json]]",
                ),
            ),
        )

        return wrapper

    return decorator


def get(url: Url = None, **kwds: a) -> Callable[[RFunc], RFunc]:
    """Get url."""

    def decorator(func: RFunc) -> RFunc:
        """Decorate function func."""

        @wraps(func)
        def wrapper(
            _url: Optional[Url] = None, *, _s: Optional[s] = None
        ) -> Res:
            """Wrap function func."""
            return func((_s or s()).get(format_url(_url or url), **kwds))

        setattr(
            wrapper,
            "__signature__",
            generate_sig(
                func,
                Parameter(
                    "_url",
                    Parameter.POSITIONAL_OR_KEYWORD,
                    default=None,
                    annotation="Optional[Url]",
                ),
                Parameter(
                    "_s",
                    Parameter.KEYWORD_ONLY,
                    default=None,
                    annotation="Optional[HTMLSession]",
                ),
            ),
        )

        return wrapper

    return decorator


def gets(urls: Urls = None, **kwds: a) -> Callable[[RsFunc], RsFunc]:
    """Get url."""

    def decorator(func: RsFunc) -> RsFunc:
        """Decorate function func."""

        @wraps(func)
        def wrapper(
            _urls: Optional[Urls] = None, *, _s: Optional[s] = None
        ) -> Reses:
            """Wrap function func."""
            return func(
                map(
                    lambda u: (_s or s()).get(format_url(u)),
                    _urls or urls or [""],
                ),
                **kwds,
            )

        setattr(
            wrapper,
            "__signature__",
            generate_sig(
                func,
                Parameter(
                    "_urls",
                    Parameter.POSITIONAL_OR_KEYWORD,
                    default=None,
                    annotation="Optional[Urls]",
                ),
                Parameter(
                    "_s",
                    Parameter.KEYWORD_ONLY,
                    default=None,
                    annotation="Optional[HTMLSession]",
                ),
            ),
        )

        return wrapper

    return decorator


def post(
    url: Optional[Url] = None,
    *,
    data: Data = None,
    json: Json = None,
    **kwds: a,
) -> Callable[[RFunc], RFunc]:
    """Post url with data and json."""

    def decorator(func: RFunc) -> RFunc:
        """Decorate function func."""

        @wraps(func)
        def wrapper(
            _url: Optional[Url] = None,
            _data: Optional[Data] = None,
            _json: Optional[Json] = None,
            *,
            _s: Optional[s] = None,
        ) -> Res:
            """Wrap function func."""
            return func(
                (_s or s()).post(format_url(url), data=data, json=json, **kwds)
            )

        setattr(
            wrapper,
            "__signature__",
            generate_sig(
                func,
                Parameter(
                    "_url",
                    Parameter.POSITIONAL_OR_KEYWORD,
                    default=None,
                    annotation="Optional[Url]",
                ),
                Parameter(
                    "_data",
                    Parameter.POSITIONAL_OR_KEYWORD,
                    default=None,
                    annotation="Optional[Data]",
                ),
                Parameter(
                    "_json",
                    Parameter.POSITIONAL_OR_KEYWORD,
                    default=None,
                    annotation="Optional[Json]",
                ),
                Parameter(
                    "_s",
                    Parameter.KEYWORD_ONLY,
                    default=None,
                    annotation="Optional[HTMLSession]",
                ),
            ),
        )

        return wrapper

    return decorator


def posts(
    urls: Optional[Urls] = None,
    *,
    datas: Optional[Iterable[Data]] = None,
    jsons: Optional[Iterable[Json]] = None,
    **kwds: a,
) -> Callable[[RsFunc], RsFunc]:
    """Post url with data and json."""

    def decorator(func: RsFunc) -> RsFunc:
        """Decorate function func."""

        @wraps(func)
        def wrapper(
            _urls: Optional[Urls] = None,
            *,
            _datas: Optional[Iterable[Data]] = None,
            _jsons: Optional[Iterable[Json]] = None,
            _s: Optional[s] = None,
        ) -> Res:
            """Wrap function func."""

            def _post(udj: Tuple[Url, Data, Json]) -> Res:
                """Make a post."""
                url, data, json = udj
                return (_s or s()).post(
                    format_url(url), data=data, json=json, **kwds
                )

            return func(
                map(
                    _post,
                    zip(
                        _urls or urls or [""],
                        _datas or datas or cycle([None]),
                        _jsons or jsons or cycle([dict()]),
                    ),
                )
            )

        setattr(
            wrapper,
            "__signature__",
            generate_sig(
                func,
                Parameter(
                    "_urls",
                    Parameter.POSITIONAL_OR_KEYWORD,
                    default=None,
                    annotation="Optional[Urls]",
                ),
                Parameter(
                    "_datas",
                    Parameter.POSITIONAL_OR_KEYWORD,
                    default=None,
                    annotation="Optional[Iterable[Data]]",
                ),
                Parameter(
                    "_jsons",
                    Parameter.POSITIONAL_OR_KEYWORD,
                    default=None,
                    annotation="Optional[Iterable[Json]]",
                ),
                Parameter(
                    "_s",
                    Parameter.KEYWORD_ONLY,
                    default=None,
                    annotation="Optional[HTMLSession]",
                ),
            ),
        )

        return wrapper

    return decorator


def css(
    selector: str = "*",
    *,
    containing: Union[str, List[str]] = None,
    clean: bool = False,
    first: bool = False,
    _encoding: str = None,
) -> Callable[[EFunc], EFunc]:
    """Css response."""

    def decorator(func: EFunc) -> EFunc:
        """Decorate function func."""

        @wraps(func)
        def wrapper(_res: Union[Res, Reses]) -> Union[Ele, Eles]:
            """Wrap function func."""
            return (
                func(
                    _res.html.find(
                        selector,
                        containing=containing,
                        clean=clean,
                        first=first,
                        _encoding=_encoding,
                    )
                )
                if isinstance(_res, Res)
                else func(
                    (lambda ele: ele if first else chain.from_iterable(ele))(
                        map(
                            lambda r: r.html.find(
                                selector,
                                containing=containing,
                                clean=clean,
                                first=first,
                                _encoding=_encoding,
                            ),
                            _res,
                        )
                    )
                )
            )

        setattr(
            wrapper,
            "__signature__",
            generate_sig(
                func,
                Parameter(
                    "_res",
                    Parameter.POSITIONAL_OR_KEYWORD,
                    default=None,
                    annotation="Res",
                ),
            ),
        )

        return wrapper

    return decorator


def xpath(
    selector: str,
    *,
    clean: bool = False,
    first: bool = False,
    _encoding: str = None,
) -> Callable[[EFunc], EFunc]:
    """Css response."""

    def decorator(func: EFunc) -> EFunc:
        """Decorate function func."""

        @wraps(func)
        def wrapper(_res: Union[Res, Reses]) -> Union[Ele, Eles]:
            """Wrap function func."""
            return func(
                _res.html.xpath(
                    selector, clean=clean, first=first, _encoding=_encoding
                )
                if isinstance(_res, Res)
                else (lambda ele: ele if first else chain.from_iterable(ele))(
                    map(
                        lambda r: r.html.xpath(
                            selector,
                            clean=clean,
                            first=first,
                            _encoding=_encoding,
                        ),
                        _res,
                    )
                )
            )

        setattr(
            wrapper,
            "__signature__",
            generate_sig(
                func,
                Parameter(
                    "_res",
                    Parameter.POSITIONAL_OR_KEYWORD,
                    default=None,
                    annotation="Res",
                ),
            ),
        )

        return wrapper

    return decorator


def html(func: SFunc) -> SFunc:
    """Get html from response."""

    @wraps(func)
    def wrapper(_res: Union[Res, Reses]) -> Union[HTML, Iterable[HTML]]:
        """Wrap function func."""
        return (
            func(_res.html)
            if isinstance(_res, Res)
            else map(lambda r: r.html, _res)
        )

    setattr(
        wrapper,
        "__signature__",
        generate_sig(
            func,
            Parameter(
                "_res",
                Parameter.POSITIONAL_OR_KEYWORD,
                default=None,
                annotation="Res",
            ),
        ),
    )

    return wrapper


def json(
    func: Callable[..., Union[Json, Iterable[Json]]]
) -> Callable[[Union[Res, Reses]], Union[Json, Iterable[Json]]]:
    """Get json from response."""

    @wraps(func)
    def wrapper(_res: Union[Res, Reses]) -> Union[Json, Iterable[Json]]:
        """Wrap function func."""
        return (
            func(_res.json())
            if isinstance(_res, Res)
            else map(lambda r: r.json(), _res)
        )

    setattr(
        wrapper,
        "__signature__",
        generate_sig(
            func,
            Parameter(
                "_res",
                Parameter.POSITIONAL_OR_KEYWORD,
                default=None,
                annotation="Res",
            ),
        ),
    )

    return wrapper


def text(func: SFunc) -> SFunc:
    """Get text from response or element(s) from css/xpath."""

    @wraps(func)
    def wrapper(_ele: Union[Res, Ele, Eles]) -> Union[str, Iterable[str]]:
        """Wrap function func."""
        return func(
            _ele.text
            if isinstance(_ele, Ele) or isinstance(_ele, Res)
            else map(lambda e: e.text, _ele)
        )

    setattr(
        wrapper,
        "__signature__",
        generate_sig(
            func,
            Parameter(
                "_ele",
                Parameter.POSITIONAL_OR_KEYWORD,
                default=None,
                annotation="Union[Res, Ele, Eles]",
            ),
        ),
    )

    return wrapper
